### PR TITLE
Last fixes before 1.23

### PR DESCRIPTION
**Pull request by jschueller**
- #2567 
- #2604 
- #2642 
- #2673
- #2674 
- #2655 
- Pagmo front computation bug when maximizing objective
- Special case for AggregatedEvaluation.getMarginal
- Fix OptimizationResult.drawOptimalValueHistory to take feasibility into account

### CODE REVIEW

**Comment by mbaudin47** [comment]
It seems that there are unexpected errors in mingw. For example:
```
terminate called after throwing an instance of 'H5::FileIException'
```

**Review by mbaudin47 — APPROVED** [review]
Thank you for fixing the Kernel Smoothing description bug.

**Review by sofianehaddad — APPROVED** [review]

**Review by regislebrun — APPROVED** [review]
Ok so the DiscreteMarkovChain class is fixed locally, not by switching to the same internal and external representation of the transition matrix. The difficulty is to manage an internal change at the save/load level.

File: lib/src/Uncertainty/Distribution/HistogramFactory.cxx [hunk 1]
```diff
@@ -109,23 +109,23 @@ Histogram HistogramFactory::buildAsHistogram(const Sample & sample,
   const Scalar mean = sample.computeMean()[0];
   if (!SpecFunc::IsNormal(mean)) throw InvalidArgumentException(HERE) << "Error: cannot build an Histogram distribution if data contains NaN or Inf";
   // It will extends from min to max.
-  const Scalar min = sample.getMin()[0];
-  const Scalar max = sample.getMax()[0];
-  if (max == min) throw InvalidArgumentException(HERE) << "Error: cannot estimate an Histogram distribution from a constant sample.";
+  const Scalar xMin = sample.getMin()[0];
+  const Scalar xMax = sample.getMax()[0];
+  if (xMax == xMin) throw InvalidArgumentException(HERE) << "Error: cannot estimate an Histogram distribution from a constant sample.";
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/HistogramFactory.cxx:114]:
```suggestion
  if (!(xMax > xMin))
    throw InvalidArgumentException(HERE) << "Error: cannot estimate a Histogram distribution from a constant sample, here max value is " << xMax << " and min value is " << xMin;
```
To be NaN-safe + typo correction

File: lib/src/Uncertainty/Distribution/HistogramFactory.cxx [hunk 2]
```diff
@@ -193,22 +193,45 @@ Scalar HistogramFactory::computeBandwidth(const Sample & sample,
     const Bool useQuantile) const
 {
   const UnsignedInteger size = sample.getSize();
-  if (size < 2) throw InvalidArgumentException(HERE) << "Error: cannot build an Histogram distribution from a sample of size < 2";
-  Scalar hOpt = 0;
+  if (size < 2)
+    throw InvalidArgumentException(HERE) << "Cannot build an Histogram distribution from a sample of size < 2";
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/HistogramFactory.cxx:197]:
```suggestion
    throw InvalidArgumentException(HERE) << "Cannot build a Histogram distribution from a sample of size < 2";
```

File: lib/src/Uncertainty/Distribution/HistogramFactory.cxx [hunk 3]
```diff
@@ -193,22 +193,45 @@ Scalar HistogramFactory::computeBandwidth(const Sample & sample,
...
+
+  const Scalar xMin = sample.getMin()[0];
+  const Scalar xMax = sample.getMax()[0];
+  if (xMin == xMax)
+    throw InvalidArgumentException(HERE) << "Cannot estimate an Histogram distribution from a constant sample.";
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/HistogramFactory.cxx:202]:
```suggestion
  if (!(xMax > xMin))
    throw InvalidArgumentException(HERE) << "Error: cannot estimate a Histogram distribution from a constant sample, here max value is " << xMax << " and min value is " << xMin;
```
To be NaN-safe + typo correction

**Review by josephmure — CHANGES_REQUESTED** [review]

File: lib/src/Uncertainty/Distribution/TrapezoidalFactory.cxx [hunk 4]
```diff
@@ -72,12 +72,11 @@ Trapezoidal TrapezoidalFactory::buildAsTrapezoidal(const Sample & sample) const
...
   Point startingPoint(4);
-  const Scalar min = sample.getMin()[0];
-  const Scalar max = sample.getMax()[0];
+  const Scalar xMin = sample.getMin()[0];
+  const Scalar xMax = sample.getMax()[0];
   const Scalar mean = sample.computeMean()[0];
   if (!SpecFunc::IsNormal(mean)) throw InvalidArgumentException(HERE) << "Error: cannot build a LogUniform distribution if data contains NaN or Inf";
-  //  if (max <= min - std::sqrt(SpecFunc::ScalarEpsilon))
-  if (min == max) throw InvalidArgumentException(HERE) << "Error: cannot estimate a Trapezoidal distribution from a constant sample.";
+  if (xMin == xMax) throw InvalidArgumentException(HERE) << "Error: cannot estimate a Trapezoidal distribution from a constant sample.";
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/TrapezoidalFactory.cxx:79]:
```suggestion
  if (!(xMax > xMin))
    throw InvalidArgumentException(HERE) << "Error: cannot estimate a Trapezoidal distribution from a constant sample, here max value is " << xMax << " and min value is " << xMin;
```
To be NaN-safe + more explicit error message

File: lib/src/Uncertainty/Distribution/HistogramFactory.cxx [hunk 5]
```diff
@@ -91,10 +91,10 @@ Histogram HistogramFactory::buildAsHistogram(const Sample & sample,
   const Scalar mean = sample.computeMean()[0];
   if (!SpecFunc::IsNormal(mean)) throw InvalidArgumentException(HERE) << "Error: cannot build an Histogram distribution if data contains NaN or Inf";
   // It will extends from min to max.
-  const Scalar min = sample.getMin()[0];
-  const Scalar max = sample.getMax()[0];
-  if (max == min) throw InvalidArgumentException(HERE) << "Error: cannot estimate an Histogram distribution from a constant sample.";
-  const UnsignedInteger binNumber = static_cast<UnsignedInteger>(ceil((max - min) / bandwidth + 0.5));
+  const Scalar xMin = sample.getMin()[0];
+  const Scalar xMax = sample.getMax()[0];
+  if (xMax == xMin) throw InvalidArgumentException(HERE) << "Error: cannot estimate an Histogram distribution from a constant sample.";
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/HistogramFactory.cxx:96]:
```suggestion
  if (!(xMax > xMin))
    throw InvalidArgumentException(HERE) << "Error: cannot estimate a Histogram distribution from a constant sample, here max value is " << xMax << " and min value is " << xMin;
```

I forgot that one.

**Review by josephmure — APPROVED** [review]
Good to go after this last change.

**Pull request merged**